### DISCUSSION TITLE
Danfoss Ally - schedule - packet publish sequence

### DISCUSSION THREAD

**Discussion by lacrimal** [Koenkk/zigbee2mqtt#14352] [Devices]
Hi,
Have in hand Danfoss Ally, want to set schedule on it, as far as I know there is no gui interface to do that so trying using Packet publishing.
sending to` zigbee2mttq/<device id>/1/set`:
This can be done with HomeAssistant as well as with Z2M Console.
Reset calendar (0x0201 0x03 1):
`{"command":{"cluster":513,"command":3,"payload":1}}`
Get callendar to be sure it was reset (0x0201 0x03 {"daystoreturn":1,"modetoreturn":1}):
`{"command":{"cluster":513,"command":2,"payload":{"daystoreturn":1,"modetoreturn":1}}}`
Send calendar payload (0x0201 0x03 <payload>)}:
`{"command":{"cluster":513,"command":1,"payload":{"dayofweek":4,"mode":1,"numoftrans":6,"transitions":[{"heatSetpoint":32,"transitionTime":0},{"heatSetpoint":32,"transitionTime":240},{"heatSetpoint":32,"transitionTime":480},{"heatSetpoint":32,"transitionTime":720},{"heatSetpoint":32,"transitionTime":960},{"heatSetpoint":32,"transitionTime":1200}]}}}`

Now it works randomly, with over 95% of failure all of them are timeouts. First of all it is EFR32 and Z2M dev, Thermostati is about 70cm from transmitter with 255 LQI.
Set command is published, sometimes it will return information (like get calendar), sometimes even calendar will be set but this is rare. 

So, should I expect that should work fine, or there is something specific that I miss ?
Is above sequence of payload right ?
Thanks

**Comment by lacrimal** [comment]
Found something interesting, update actually works but output payload is weirdly updated - you will not see schedule data but only it influence on other parameters for example:
<pre>
....
    <b>"occupied_heating_setpoint": 10.3,</b>
    "pi_heating_demand": 0,
    "preheat_status": false,
    "programming_operation_mode": "schedule",
    "radiator_covered": false,
    "running_state": "idle",
    <b>"setpoint_change_source": "manual",</b>
    "system_mode": "heat",
    "trigger_time": 241,
    "update": {
        "state": "idle"
    },
    "weekly_schedule": {
        "days": [
            "sunday"
        ],
        "transitions": []
    },
....
</pre>
after disposing payload (over HomeAssistant MQTT add-on but same effect is with sending payload to cluster 512 1 with Z2M console per device):
`{"command":{"cluster":513,"command":1,"payload":{"dayofweek":8,"mode":1,"numoftrans":6,"transitions":[{"heatSetpoint":1220,"transitionTime":0},{"heatSetpoint":1220,"transitionTime":240},{"heatSetpoint":1220,"transitionTime":480},{"heatSetpoint":1220,"transitionTime":720},{"heatSetpoint":1220,"transitionTime":960},{"heatSetpoint":1220,"transitionTime":1200}]}}}`

we get:

<pre>
    <b>"occupied_heating_setpoint": 12.2,</b>
    "pi_heating_demand": 0,
    "preheat_status": false,
    "programming_operation_mode": "schedule",
    "radiator_covered": false,
    "running_state": "idle",
    <b>"setpoint_change_source": "schedule",</b>
    "system_mode": "heat",
    "trigger_time": 241,
    "update": {
        "state": "idle"
    },
    "weekly_schedule": {
        "days": [
            "sunday"
        ],
        "transitions": []
    },
</pre>

But no change in payload considering weekly_schedule, also internal time is never synchronised as it set as default 0x2000E3B0 (Jan 5th
2017, 11:00 AM ), as documentation said :_The default value is synchronized at boot where the side MCU sends it in DATETIME format and the top ZigBee MCU converts it to UTC_ but that not gonna happen.

So more questions - can issue with time synchronisation as well as for schedule presentation can be fixed ?
I will remind I'm on EFR32 with Z2M dev.

**Comment by akuzm** [comment]
Facing the same issues on Zigbee2MQTT version 1.33.0 (commit #4d53ac4) and zigbee-herdsman (0.18.5).

For the reference, the message to reset the internal time of thermostat is `{"write": {"cluster": "genTime", "payload": {"time": timestamp}}},` where `timestamp` is the number of seconds since 1 Jan 2000 UTC, or, in Python, `(datetime.datetime.now().astimezone() - dateutil.parser.parse("01 Jan 2000 00:00:00+0000")).total_seconds())`. It must be sent to the same `zigbee2mttq/<device id>/1/set` queue.

**Comment by aswarcewicz** [comment]
Hello, maybe somebody of you can help. I reset valve to factory settings. Set programming mode to schedule, then set schedule. And no changes, I try move trigger_time property, try to wait for time change, nothing changes…. Schedule set for all days ( value 127)

**Comment by lacrimal** [comment]
Is schedule really set?
Did you listed all days to verify ?
If yes - did you set your time also ?
If you are using node red - I have workflow to ser time and schedule.

**Reply by aswarcewicz** [reply]
After schedule set response returns success. I will try to get every day and check. After factory reset time is reset also and I do not set it. This the reason why I set schedule for all days and different times, now I only try to get it working

**Reply by aswarcewicz** [reply]
Ok I check it, there is no response when I try to get schedule for specific day
![image](https://github.com/Koenkk/zigbee2mqtt/assets/3719103/8de7e312-6230-41a5-bc18-82b774027cbf)

**Reply by lacrimal** [reply]
You need to query specific day then go to 'State' tab. For danfoss there is no better mechanism implemented.

**Reply by lacrimal** [reply]
Time in danfoss I counted  from 2000-01-01 00:00, try any online calculator to check of you are close.
You need to set time individually for your danfoss, controller with software should do it but I never get it to work, or get sufficient help on forum so I made it with node red.
Mine time with timezones is  742120819 (pretty sure we are in same time zone:-)) so your looks wrong, like manufacturer state. If your time zone shifts and you are in DST you need also set other parameters ow genTime.
Schedule should be similar to this in your state tab:
   "weekly_schedule": {
        "days": [
            "saturday"
        ],
        "transitions": [
            {
                "heating_setpoint": 18,
                "time": 450
            },
            {
                "heating_setpoint": 22,
                "time": 1110
            }
        ]
    },

**Reply by aswarcewicz** [reply]
Thanks for quick answers!
So my schedule do not apply.
I try to set:
`{
    "dayofweek":64,
    "mode":1,
    "numoftrans":6,
    "transitions": [
      {"heatSetpoint":10,"transitionTime":0},
      {"heatSetpoint":11,"transitionTime":200},
      {"heatSetpoint":12,"transitionTime":300},
      {"heatSetpoint":15,"transitionTime":500},
      {"heatSetpoint":16,"transitionTime":700},
      {"heatSetpoint":19,"transitionTime":1000},
      {"heatSetpoint":20,"transitionTime":1400}
    ]
}`
Then when i get schedule for saturday:
`{"daystoreturn":64,"modetoreturn":1}`
It is simply empty...
`{
    "abs_max_heat_setpoint_limit": 35,
    "abs_min_heat_setpoint_limit": 5,
    "adaptation_run_control": "none",
    "adaptation_run_settings": true,
    "adaptation_run_status": "none",
    "algorithm_scale_factor": 1,
    "battery": 59,
    "day_of_week": "wednesday",
    "external_measured_room_sensor": -8000,
    "heat_available": true,
    "heat_required": false,
    "keypad_lockout": "unlock",
    "last_seen": "2023-10-14T09:16:48.894Z",
    "linkquality": 65,
    "load_balancing_enable": false,
    "load_estimate": -8000,
    "load_room_mean": -8000,
    "local_temperature": 21.48,
    "local_temperature_calibration": null,
    "max_heat_setpoint_limit": 35,
    "mounted_mode_active": false,
    "mounted_mode_control": false,
    "occupancy": false,
    "occupied_heating_setpoint": 15.5,
    "occupied_heating_setpoint_scheduled": 15.5,
    "pi_heating_demand": 0,
    "preheat_status": false,
    "programming_operation_mode": "schedule",
    "radiator_covered": false,
    "regulation_setpoint_offset": 0,
    "running_state": "idle",
    "setpoint_change_amount": null,
    "setpoint_change_source": "manual",
    "system_mode": "heat",
    "temperature_setpoint_hold": false,
    "thermostat_vertical_orientation": false,
    "trigger_time": 0,
    "unoccupied_heating_setpoint": null,
    "weekly_schedule": {
        "days": [
            "saturday"
        ],
        "transitions": []
    },
    "window_open_external": false,
    "window_open_feature": false,
    "window_open_internal": "closed",
    "update": {
        "installed_version": -1,
        "latest_version": -1,
        "state": null
    },
    "update_available": null,
    "viewing_direction": null
}`

I do not know why?

**Reply by lacrimal** [reply]
You have 7 transitions but declared 6, back in time this also take part of my life to figure out.
Anyway - I'm using raw command to cluster 513, try it , maybe change days for all and then query any day to see if it is applied:
`{
    "command": {
        "cluster": 513,
        "command": 1,
        "payload": {
            "dayofweek": 127,
            "mode": 1,
            "numoftrans": 2,
            "transitions": [
                {
                    "heatSetpoint": 2400,
                    "transitionTime": 390
                },
                {
                    "heatSetpoint": 2100,
                    "transitionTime": 1320
                }
            ]
        }
    }
}`

Next you will need to set right time - like I said - to have everything under control - I'm using nodered.

**Comment by aswarcewicz** [comment]
For setting time I make simple JavaScript snippet. But I hope it is not required for testing and I need it to correct apply temperature based on schedule? If yes it is my next step.
My JS snippet for Danfoss time:

```
var nowForSetInAlly = (new Date().getTime() - Date.parse("01 Jan 2000 00:00:00+0000")) / 1000;


// last set time after factory reset was 4294967295
var timeFromAlly = 4294967295;
var timeFromAllyAsString = new Date(timeFromAlly * 1000 + Date.parse("01 Jan 2000 00:00:00+0000"));


console.log("nowForSetInAlly: " + nowForSetInAlly);
console.log("timeFromAllyAsString: " + timeFromAllyAsString);
console.log("Diff: " + (nowForSetInAlly - timeFromAllyAsString))
```

**Reply by lacrimal** [reply]
To test your time is correctly applied you can use combination of 'day_of_week' and 'trigger_time' it will trigger head full open and close at defined time of day.

**Reply by aswarcewicz** [reply]
Ok it is clear. But now I know that my schedule is not apply. I do not know reason yet...
I again try factory reset and apply schedule after it but nothing changes. After query for schedule it is still empty

**Reply by lacrimal** [reply]
Like I said - you have error in numoftrans.
Try apply mine schedule - it was tested and works.

**Reply by aswarcewicz** [reply]
Thank you very much now it looks much better!
Sorry for delay but in a meanwhile I upgrade firmware of my coordinator - now everything seems better. Previous I have a problem with window open function now it is correctly updated

**Reply by lacrimal** [reply]
Great,I have EFR32, so still experimental.

**Reply by aswarcewicz** [reply]
@lacrimal I few times do factory reset and now I see that probably zigbee2mqtt set time in Danfoss TRV:
`Info 2023-10-14 13:17:41Read result of 'genTime': {"time":750597459,"timeZone":7200,"localTime":750604659,"lastSetTime":750593527}
`
Do you observe similar behaviour?
I only need to check is localTime correct - I will verify It comparing to schedule
edit: time seems correct

**Comment by polymeris** [comment]
FWIW, this thread helped me get the schedule working. Here is the relevant snippet of code (using AppDaemon):

``` python3
class DanfossTRV(Thermostat):
  """ Implements Danfoss Ally TRV specifics.
  Assumes TRV Programming operation mode = "schedule" or "schedule_with_preheat"
  Pending:
  - Toggle Heat available depending on season
  - External window open detection
  - Load balancing when multiple TRVs in one room
  """

  def reconfigure(self):
    """ Called when state of config entities changes """
    self.adapi.log(f"Reconfiguring {self.entity.friendly_name}")
    self._sync_time()
    self._set_schedule()

  def _set_schedule(self):
    """ Set the TRV schedule based on self.transitions
    which is a list of objects containing fields:
      temperature: float
      start: time
    """
    if not self.transitions:
      return
    
    self.adapi.log(f"Configuring transitions {self.transitions}")
    # c.f. https://assets.danfoss.com/documents/176987/AM375549618098en-000101.pdf
    schedule_msg = {
      "command": {
        "cluster": 0x0201,
        "command": 0x01,
        "payload": {
            # days as bits  SFTWTMS
            "dayofweek": 0b01111111,
            "mode": 1, # 1 = heat
            "numoftrans": len(self.transitions),
            "transitions": [
              {
                # minutes from midnight
                "transitionTime": t.start.hour * 60 + t.start.minute,
                # temp seems to be expected in milli-celsius, even tho it's reported back in C (with command 0x02)
                "heatSetpoint": int(t.temperature * 100)
              }
              for t in self.transitions]
          }
      }
    }

    self._mqtt_set(schedule_msg)
  
  def _sync_time(self):
    """ Updates clock on TRV (seconds since January 1st, 2000) """
    import time
    time_msg = {
      "write": { 
        "cluster": "genTime",
        "payload": {
            "time": int(datetime.now().timestamp() - datetime(2000, 1, 1).timestamp()),
            "timeStatus": 0b10,
            "timeZone": -time.timezone
          }
      }
    }
    self._mqtt_set(time_msg)

  def _mqtt_set(self, payload):
    self.adapi.call_service("mqtt/publish", topic = f"zigbee2mqtt/{self.entity.friendly_name}/1/set", payload=json.dumps(payload))
```
